include/xtensor/xfunction.hpp

```diff
@@ -1012,13 +1012,17 @@ namespace xt
     template <class F, class R, class... CT>
     inline bool xfunction_iterator<F, R, CT...>::equal(const self_type& rhs) const
     {
-        return p_f == rhs.p_f && m_it == rhs.m_it;
+        // Optimization: no need to compare each subiterator since they all
+        // are incremented decremented together.
+        return p_f == rhs.p_f && std::get<0>(m_it) == std::get<0>(rhs.m_it);
```

Comment:
IMO we can also remove the `p_f == rhs.p_f` because ... it's your problem if you compare the wrong iterators and it should segfault. (otherwise you get even weirder results... like no iteration at all)

Reply:
Yep I wondered if I should removed it, because the equality operator should return true if the compared elements are the same, and false if they're different, otherwise it's semantically wrong. I know we did that elsewhere but I still wonder if it was a good idea.

Reply:
Finally the holy standard gives the answer:

§ 24.2.1
An iterator j is called reachable from an iterator i if and only if there is a finite sequence of applications of the expression ++i that makes i == j. If j is reachable from i, they refer to elements of the same sequence.

§ 24.2.5
The domain of == for forward iterators is that of iterators over the same underlying sequence.

Thus comparing iterators from different containers is undefined.

Reply:
By the way, a mean to guarantee that we don't need to compare all the subiterators would be to make the constructor of `xfunction_iterator` private, and declare it as a friend class in `xfunction`.